### PR TITLE
[terra-application] Reads out error boundary message

### LINKED ISSUE #56: Investigate our options for focus with terra-application-error-boundary
# Feature Request

Currently, it is indeterminate what happens to the active element when the error boundary shows (in chrome and firefox it goes to body, in ie it clears it).  We should figure out if there's something we can do better here.

### PR DESCRIPTION
**Pull request by lokesh-0813**
### Summary

Allows error boundary to read out error messages on screen readers.
~Focus is set on error boundary once rendered. ~

Closes #56 

### Deployment Link

https://terra-applic-.herokuapp.com/

### Testing

Tested in Chrome/Jaws and Safari/Voicover
### Additional Details

Thank you for contributing to Terra.
@cerner/terra

### CODE REVIEW

File: src/application-error-boundary/ApplicationErrorBoundary.jsx [hunk 1]
```diff
@@ -92,10 +95,12 @@ class ApplicationErrorBoundary extends React.Component {
       const errorDetails = activeError.message.toString();
       const errorText = intl.formatMessage({ id: 'terraApplication.errorBoundary.defaultErrorMessage' }, { errorDetails });
       return (
-        <StatusView
-          variant="error"
-          message={errorText}
-        />
+        <div role="alert" id="applicationErrorBoundary" tabIndex="-1">
```
Comment (tbiethman) [inline src/application-error-boundary/ApplicationErrorBoundary.jsx]: An `id` is probably safe, but using a `ref` would be more safe.

Comment (lokesh-0813) [reply src/application-error-boundary/ApplicationErrorBoundary.jsx]: Switched to `ref`.

**Comment by lokesh-0813** [comment]
With the current changes the browser adds a default focus-border around the application-error-boundary. Should we consider removing the border styles or wait till UX provides a value?

File: src/application-error-boundary/ApplicationErrorBoundary.jsx [hunk 2]
```diff
@@ -92,10 +96,12 @@ class ApplicationErrorBoundary extends React.Component {
       const errorDetails = activeError.message.toString();
       const errorText = intl.formatMessage({ id: 'terraApplication.errorBoundary.defaultErrorMessage' }, { errorDetails });
       return (
-        <StatusView
-          variant="error"
-          message={errorText}
-        />
+        <div ref={this.errorBoundaryRef} role="alert" tabIndex="-1">
```
Comment (tbiethman) [inline src/application-error-boundary/ApplicationErrorBoundary.jsx]: Is the "alert" role is appropriate here? It does ensure that the error text is "assertive" and that it's read when presented, but I'm not sure if there are other side effects.

File: src/application-error-boundary/ApplicationErrorBoundary.jsx [hunk 3]
```diff
@@ -67,6 +67,10 @@ class ApplicationErrorBoundary extends React.Component {
 
   componentDidUpdate() {
     this.resetError();
+
+    if (this.errorBoundaryRef.current) {
+      this.errorBoundaryRef.current.focus();
```
Comment (tbiethman) [inline src/application-error-boundary/ApplicationErrorBoundary.jsx]: Moving focus on every update is likely too often. The error boundary may be visible while the user interacts with other parts of the app (like the navigation header), and we don't want to move focus when those updates trickle down to the error boundary.

Can we limit this focus logic to just the initial presentation? I think _ideally_ we'd only move focus if the item that was previously focused is removed as part of the status view presentation, but I don't know if we can determine that reliably.

Comment (tbiethman) [reply src/application-error-boundary/ApplicationErrorBoundary.jsx]: Thinking about this some more, I'm not sure if we need to be moving focus at all. Using the "alert" role might be enough to ensure the error is read and accessible. Setting focus to the error may make sense if the previously focused element was removed during the error presentation, but if not, the focus moving may be more obtrusive than we expect.

File: src/application-error-boundary/ApplicationErrorBoundary.jsx (hunk 2 above)
Comment (ShettyAkarsh) [inline src/application-error-boundary/ApplicationErrorBoundary.jsx]: Wrapping with div would add a default height. which would avoid error status to be displayed in the centre of the page when page level error boundary is implemented.  we might have to set the height of this div to 100% in that case.

File: src/application-error-boundary/ApplicationErrorBoundary.jsx (hunk 3 above)
Comment (lokesh-0813) [reply src/application-error-boundary/ApplicationErrorBoundary.jsx]: Removed focus in this commit 1b2a937569096398257177107586295f33a1a467

Comment (lokesh-0813) [reply src/application-error-boundary/ApplicationErrorBoundary.jsx]: > Thinking about this some more, I'm not sure if we need to be moving focus at all. Using the "alert" role might be enough to ensure the error is read and accessible. Setting focus to the error may make sense if the previously focused element was removed during the error presentation, but if not, the focus moving may be more obtrusive than we expect.

That makes sense too.

File: src/application-error-boundary/ApplicationErrorBoundary.jsx (hunk 2 above)
Comment (lokesh-0813) [reply src/application-error-boundary/ApplicationErrorBoundary.jsx]: Removed the `DIV` and role is now a `customProp`

Changed here 1b2a937569096398257177107586295f33a1a467

Comment (ShettyAkarsh) [reply src/application-error-boundary/ApplicationErrorBoundary.jsx]: I am not sure. but i had question like do we need to add this role attribute to status-view itself ? there might be this same accessibility issue in applications using status view. so which they have to fix it by passing the role attribute again similar to this. 

@tbiethman , @ryanthemanuel , @lokesh-0813  your thoughts on this  ?

**Review by tbiethman — APPROVED** [review]

**Review by mjhenkes — APPROVED** [review]

**Review by jeremyfuksa — APPROVED** [review]

**Pull request merged**